Handle tail 0-size tensor appropriately in `MultiTensorApply`

crcrpar opened the pull request:
Fixes #100701

It seems like we don't call `multi_tensor_apply_kernel` at all if the input tensor lists are small and their last tensors are zero-size as per e.g. https://github.com/pytorch/pytorch/blob/ca9f55f79d944672cb93157836f8ee92f54d2e10/aten/src/ATen/native/cuda/MultiTensorApply.cuh#L100-L102
which was introduced in https://github.com/pytorch/pytorch/commit/05943712a443138497c185405b575043b2916f34.

This PR special cases the last zero-size tensors so that we won't be negligent.

cc @ngimel

crcrpar committed:
disable MultiTensorApply if the last Tensor is 0-size

also adds test for foreach optimizers

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
add tests to fused optimizers

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
handle zero-size tensors better in multi_tensor_apply

Basically the core change is that if the tail tensor is zero-size,
skip `continue` and goes to the multi_tensor_apply_kernel caller so that
we won't miss necessary cuda kernel calls.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
revert changes in error message

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

awgu commented:
Ah, nice PR! Sorry about the bug :(

crcrpar commented:
@pytorchbot label ciflow/trunk

ngimel commented:
btw, should there be a test for regular for_each ops? Usually those are easier to debug than optimizer tests, and imo tests should be the smallest piece of code demonstrating the problem.

crcrpar committed:
skip the foreach optim test on rocm

I have no rocm devices but it feels like it somehow unexpectedly
skip `multi_tensor_apply_kernel` call.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
> Ah, nice PR! Sorry about the bug :(

cheers, zero size tensors are complicated, I know

> should there be a test for regular for_each ops?

agreed but test_foreach is not that easy to modify (and I think I've told you something like I'll refactor it shortly a few times so far but 😓)

janeyx99 commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -152,32 +161,40 @@ void multi_tensor_apply(
-            for (int d = 0; d < depth; d++) {
-                tensorListMeta.addresses[d][loc_tensor_info] = tensor_lists[d][t].const_data_ptr();
+                n_zero_tensors++;
+                if (t != n_tensors - 1) {
```
I see this is built on top of existing code, but the various "t == n_tensors - 1" equality checks (in 174, 188, and 197) are not clean. I'm thinking we should be able to break this code into a for loop iterating through tensors that update the n_zero_tensors, tensorListMeta and chunk information followed by the `multi_tensor_apply_kernel` (which is only run on the last chunk anyway, so it really should not live in the loop).

crcrpar replied:
I'm not sure if that will be completely equivalent from what I understand

crcrpar replied:
I haven't been good enough to do that. If you've got something working already, feel free to take over this

ngimel replied:
If it's a hard refactor, we should land as is, because it fixes a major bug, sorry, I don't have anything working.

janeyx99 replied:
Sounds reasonable to me, I'll file an issue to just track that this as a valuable next step. https://github.com/pytorch/pytorch/issues/101096

janeyx99 commented on test/test_optim.py:
```diff
@@ -820,6 +866,10 @@ def test_multi_tensor_optimizers(self):
         self._test_derived_optimizers(optimizer_pairs_with_flags, "foreach")
+        # TODO(crcrpar): Investigate ROCm failure in
+        # https://github.com/pytorch/pytorch/actions/runs/4910476500/jobs/8767913127
+        if not TEST_WITH_ROCM:
```
This should be investigated before landing the PR

janeyx99 replied:
Even we do not have access to ROCm machines, as the CI is provided from AMD 😛. 

The first thing I would try here is seeing if adding an empty param to the param list in _test_derived_optimizers (the merge I recommend) would still fail ROCm. If it still does, we can involve the AMD people on Slack to get help debugging.

janeyx99 replied:
Also, this issue is likely relevant just FYI https://github.com/pytorch/pytorch/issues/69698.

I'd still recommend merging with the other test.

janeyx99 commented on test/test_optim.py:
```diff
@@ -779,6 +780,51 @@ def _test_derived_optimizers(self, optimizer_pairs_with_flags, flag):
                     self.assertEqual(st_p_state[k], mt_p_state[k])
 
+    # ref: https://github.com/pytorch/pytorch/issues/100701
+    def _test_derived_optimizers_with_zero_size_params(self, optimizer_pairs_with_flags, flag):
```
This should be merged with _test_derived_optimizers by adding a zero tensor to the params of the model there, since we know the foreach=False works as a good reference.

crcrpar commented on aten/src/ATen/native/cuda/MultiTensorApply.cuh:
```diff
@@ -4,6 +4,7 @@
 #include <c10/cuda/CUDAGuard.h>
 #include <ATen/native/cuda/Loops.cuh>
 #include <ATen/native/cuda/MemoryAccess.cuh>
+#include <vector>
```
"Include What You Use" stuff

janeyx99 replied:
Does this refer to the same thing as std::vector?

crcrpar replied:
yes

ngimel commented:
When generic test is not easy to modify, there's no law against creating a one-off test testing exactly the thing you need. Use this exception sparingly, but it exists.

janeyx99 commented on test/test_optim.py:
```diff
@@ -748,8 +748,15 @@ def _test_derived_optimizers(self, optimizer_pairs_with_flags, flag):
+                # ref: https://github.com/pytorch/pytorch/issues/100701
+                # TODO(crcrpar): Investigate ROCm failure in
+                # https://github.com/pytorch/pytorch/actions/runs/4910476500/jobs/8767913127
+                empty_params = [torch.empty((), device=device, dtype=torch.float64)] if flag_value else []
```
Is flag_value check necessary here? It should still pass even if we just always added an empty param even in the non foreach/fused case, no? One would argue it might be more important to test the empty param in all fused/foreach/forloop.
```suggestion
                empty_params = [torch.empty((), device=device, dtype=torch.float64)]
```

janeyx99 commented on test/test_optim.py:
```diff
@@ -748,8 +748,15 @@ def _test_derived_optimizers(self, optimizer_pairs_with_flags, flag):
+                # foreach/fused optimizers should be tested with a param_groups['params'] with
+                # zeor_size tensor as its last param.
+                # ref: https://github.com/pytorch/pytorch/issues/100701
+                # TODO(crcrpar): Investigate ROCm failure in
+                # https://github.com/pytorch/pytorch/actions/runs/4910476500/jobs/8767913127
```
```suggestion
```
This passes in CI, right?

crcrpar commented:
@pytorchbot merge

crcrpar commented:
giving another shot as GitHub Status says the incident has been resolved.

@pytorchbot merge

crcrpar commented:
@pytorchbot merge -r

crcrpar commented:
@pytorchbot merge
